backend/src/lib/server-side-events.ts

```diff
@@ -0,0 +1,277 @@
+        try {
+            if (client.compressionStream) client.compressionStream.flush()
+        } catch (err) {
+            logger.error(err)
+        }
+        client.processing = false
+    }
+
+    private static createEventString(event: IEvent): string {
+        let eventString = `id:${event.id}\n`
+        if (event.name) eventString += `event:${event.name}\n`
+        switch (typeof event.data) {
+            case 'string':
+            case 'number':
+            case 'bigint':
+                eventString += `data:${event.data}\n`
+                break
+            case 'boolean':
+                eventString += `data:${event.data ? 'true' : 'false'}\n`
+                break
+            case 'object':
+                try {
+                    eventString += `data:${JSON.stringify(event.data)}\n`
+                } catch (err) {
+                    logger.error(err)
+                    return
+                }
+                break
+            default:
+                return
+        }
+        eventString += '\n'
+        return eventString
+    }
+
+    public static removeEvent(eventID: number): void {
+        delete this.events[eventID]
+    }
+
+    public static handleRequest(clientID: string, req: Http2ServerRequest, res: Http2ServerResponse): IEventClient {
+        const [writableStream, compressionStream, encoding] = getEncodeStream(
+            (res as unknown) as NodeJS.WritableStream,
+            req.headers[HTTP2_HEADER_ACCEPT_ENCODING] as string
+        )
+
+        let events: Record<string, boolean>
+        let namespaces: Record<string, boolean>
+        const queryStringIndex = req.url.indexOf('?')
+        if (queryStringIndex !== -1) {
+            const queryString = req.url.substr(queryStringIndex + 1)
+            const parts = queryString.split('&')
+            for (const part of parts) {
+                if (part.startsWith('events=')) {
+                    events = part
+                        .substr(7)
+                        .split(',')
+                        .reduce((events, event) => {
+                            events[event] = true
+                            return events
+                        }, {} as Record<string, boolean>)
+                } else if (part.startsWith('namespaces=')) {
+                    namespaces = part
+                        .substr(11)
+                        .split(',')
+                        .reduce((namespaces, namespace) => {
+                            namespaces[namespace] = true
+                            return namespaces
+                        }, {} as Record<string, boolean>)
+                }
+            }
+        }
+        const eventClient: IEventClient = {
+            events,
+            namespaces,
+            writableStream,
+            compressionStream,
+            eventQueue: [],
+        }
+        this.clients[clientID] = eventClient
+
+        res.setTimeout(2147483647)
+        req.setTimeout(2147483647)
```

Comment:
super large timeout. I assume it's for sse & http2? do we need to modify the ingress? 

I remember for websocket if 30 seconds no messages, will force close the connection. Not sure if the SSE having the problem if we are not polling anymore?

Reply:
nvm, I saw the keep alive, but seems it's 60 seconds?